### PR TITLE
Add GitHub syncing infrastructure

### PR DESCRIPTION
**Pull request by mrobinson**
This is a replacement for the Gecko-based sync for WebRender. Instead of having a remote task create a pull request with upstream changes, this adds an action which:

1. Clones a copy of the git mirror of gecko-dev.
2. Filters that repository and adds the filtered repository as a remote.
3. Cherry-picks new commits based off a file which keeps a hash of the last cherry-picked commit.
4. Pushes the new contents to `main` in this repository as well as a commit that updates the hash.

Landing this change will also require renaming the `master` branch to `main`.

### CODE REVIEW

File: .github/sync/sync.sh [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+step "Filtering upstream"
+# Cloning and filtering is much faster than git filter-repo --source --target.
+git clone upstream -- "$filtered"
+git -C "$filtered" filter-repo --force --paths-from-file "$ROOT/.github/sync/webrender.paths"
+
+step "Adding filtered repository as a remote"
+cd "$ROOT"
+git remote add upstream "$filtered"
+git fetch upstream
```
Comment (delan) [inline .github/sync/sync.sh:58]: rename the remote to `filtered` to avoid confusing it with the raw upstream above

File: .github/sync/sync.sh [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+hash_file=".github/sync/UPSTREAM_COMMIT"
+hash=`cat $hash_file`
+number_of_commits=`git log $hash..upstream/master --pretty=oneline | wc -l`
+
+if [ $number_of_commits != '0' ]; then
+    step "Applying $number_of_commits new commits"
+    git cherry-pick $hash..upstream/master
```
Comment (delan) [inline .github/sync/sync.sh:66]: do we have a plan for what happens when the filtered hashes change? will this work?

File: .github/sync/sync.sh [hunk 3]
```diff
@@ -0,0 +1,74 @@
+#!/bin/sh
+# Usage: sync.sh <path/to/filtered>
+set -eu
+
+DOWNSTREAM="git@github.com:mrobinson/webrender.git"
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply .github/sync/sync.sh]: I've removed this.

File: .github/sync/sync.sh [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ROOT=$(pwd)
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply .github/sync/sync.sh]: Renamed to `root_dir` for consistency with `cache_dir`.

File: .github/sync/sync.sh [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+CACHE_DIR=$ROOT/_cache
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply .github/sync/sync.sh]: Oh yeah, good point. I've been writing too much Python, I guess. :sweat_smile:

File: .github/sync/sync.sh [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+# Configure git because we will be making commits.
+git config --global user.email "noreply@github.com"
+git config --global user.name "Webrender Upstream Sync"
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply .github/sync/sync.sh]: Sure. I've switched to this approach.

File: .github/sync/sync.sh [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    if [ "${TERM-}" != '' ]; then
+        tput setaf 12
+    fi
+    >&2 printf '* %s\n' "$*"
+    if [ "${TERM-}" != '' ]; then
+        tput sgr0
+    fi
+}
+
+step "Creating work directory for sync"
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply .github/sync/sync.sh]: Done.

File: .github/sync/sync.sh [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+mkdir -p "$1"
+cd -- "$1"
+filtered=$(pwd)
+
+step "Creating cache diretory if needed"
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply .github/sync/sync.sh]: Fixed this!

File: .github/sync/sync.sh (hunk 1 above)
Comment (mrobinson) [reply .github/sync/sync.sh]: Good point. I've gone with `filtered-upstream` just to be even clearer.

File: .github/sync/sync.sh (hunk 2 above)
Comment (mrobinson) [reply .github/sync/sync.sh]: I've added a bunch of documentation to the top of this file which I hope covers that case.

File: .github/sync/sync.sh [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+number_of_commits=`git log $hash..upstream/master --pretty=oneline | wc -l`
+
+if [ $number_of_commits != '0' ]; then
+    step "Applying $number_of_commits new commits"
+    git cherry-pick $hash..upstream/master
+    git rev-parse upstream/master > "$hash_file"
+    git commit "$hash_file" -m "Syncing to upstream (`cat $hash_file`)"
+
+    step "Pushing new main branch"
+    git push origin main
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply .github/sync/sync.sh]: Yep, that makes total sense. Done.

**Review by mrobinson — COMMENTED** [review]
Thanks for the review, @delan. I've added a bunch of documentation to the top of `sync.sh` for how this script works and what to do when the hash changes. I don't think it makes sense to add any documentation to the root, because I want to avoid the possibility of conflicts when cherry-picking new commits.

File: .github/sync/sync.sh (hunk 2 above)
Comment (delan) [reply .github/sync/sync.sh]: looks great, thanks!

**Review by delan — APPROVED** [review]
Thanks for the quick update! I feel a bit iffy about putting our own files in .github, maybe we want to move .github/sync to .sync or .servo/sync or something? Otherwise this patch looks good to go.

> I don't think it makes sense to add any documentation to the root, because I want to avoid the possibility of conflicts when cherry-picking new commits.

That’s fair, I guess the README does explain that this is a fork at least.

**Pull request merged**